src/rdkafka.h

```diff
@@ -2270,6 +2284,60 @@ typedef enum rd_kafka_cert_enc_t {
 } rd_kafka_cert_enc_t;
 
 
+/**
+ * @brief Sets a callback that provides a client certificate
+ *
+ * This callback, if set, is triggered from internal librdkafka threads
+ * upon connecting to a broker. If it returns a TLS certificate/key,
+ * this is used in preference to any value set with the `ssl.key.` and
+ * `ssl.certificate.` family of options.
+ *
+ * The pointers *certificate_buf_size and *key_buf_size initially contain the
+ * maximum amount of bytes that can be written to *certificate_buf and *key_buf;
+ * the callback is expected to overwrite these values with the actual length of
+ * the certificate/key data it wrote (which should be no larger than the initial
+ * value of the buffer size)
+ *
+ * If the callback is able to provide a client certificate/key to librdkafka,
+ * it should:
+ *   - Return RD_KAFKA_CERT_FETCH_OK,
+ *   - write a certificate to certificate_buf,
+ *   - write the format of that certificate to certificate_format,m
+ *   - write the size of the certificate to certificate_buf_size,
+ *   - write a key to key_buf,
+ *   - write the format of that key to key_format
+ *   - write the size of the key to key_buf_size,
+ *
+ * Otherwise, if the provided buffers are too small (i.e *certificate_buf_size
+ * or *key_buf_size are too small), then the callback should:
+ *   - Return RD_KAFKA_CERT_FETCH_MORE_BUFFER,
+ *   - write the desired buffer size to *buf_size
+ * Note that the first time rd_kafka_conf_set_ssl_cert_fetch_cb is called, it is
```

Comment:
Can we flip this around and simply have the callback allocate the memory (through rd_kafka_malloc) and have librdkafka free it when done?

Reply:
Yeah I'm not really a C expert but I thought it would be "ugly" to transfer ownership of the buffers like that as part of the callback interface. It would work though and make the API probably simpler to use. So I'll make this change, and document it appropriately.

Reply:
Hum. So I went looking at this, there _is_ no `rd_kafka_malloc`. I'm going to define `rd_kafka_mem_alloc` as an analogue to `rd_kafka_mem_free` and have it just wrap `rd_malloc`.

Reply:
🤦  so there _wasn't_ an allocator function on my branch... so I implemented one as `rd_kafka_mem_alloc`... then rebased onto master to fixup conflicts, and lo and behold, `rd_kafka_mem_malloc` appears. Now I feel daft.